crates/shared/src/price_estimation/sanitized.rs

```diff
@@ -0,0 +1,331 @@
+    pub fn new(
+        inner: T,
+        native_token: H160,
+        bad_token_detector: Arc<dyn BadTokenDetecting>,
+    ) -> Self {
+        Self {
+            inner,
+            native_token,
+            bad_token_detector,
+        }
+    }
+
+    async fn get_token_quality_errors(
+        &self,
+        queries: &[Query],
+    ) -> HashMap<H160, PriceEstimationError> {
+        let mut token_quality_errors: HashMap<H160, PriceEstimationError> = Default::default();
+        let mut checked_tokens = HashSet::<H160>::default();
+
+        for token in queries
+            .iter()
+            .copied()
+            .flat_map(|query| [query.buy_token, query.sell_token])
+        {
+            if checked_tokens.contains(&token) {
+                continue;
+            }
+
+            match self.bad_token_detector.detect(token).await {
+                Err(err) => {
+                    token_quality_errors.insert(token, PriceEstimationError::Other(err));
+                }
+                Ok(TokenQuality::Bad { .. }) => {
+                    token_quality_errors
+                        .insert(token, PriceEstimationError::UnsupportedToken(token));
+                }
+                _ => (),
+            };
+            checked_tokens.insert(token);
+        }
+        token_quality_errors
+    }
+
+    async fn bulk_estimate_prices<'a, 'b>(
+        &self,
+        queries: &'a [Query],
+    ) -> (Vec<EstimationProgress<'b>>, Vec<EstimationResult>)
+    where
+        'a: 'b,
+    {
+        let token_quality_errors = self.get_token_quality_errors(queries).await;
+
+        let mut estimations_to_forward = Vec::new();
+
+        let estimation_progress = queries
+            .iter()
+            .map(|query| {
+                if let Some(err) = token_quality_errors.get(&query.buy_token) {
+                    return EstimationProgress::TrivialSolution(Err(err.clone()));
+                }
+                if let Some(err) = token_quality_errors.get(&query.sell_token) {
+                    return EstimationProgress::TrivialSolution(Err(err.clone()));
+                }
+
+                if query.buy_token == query.sell_token {
+                    let estimation = Estimate {
+                        out_amount: query.in_amount,
+                        gas: 0.into(),
+                    };
+
+                    tracing::debug!(?query, ?estimation, "generate trivial price estimation");
+                    return EstimationProgress::TrivialSolution(Ok(estimation));
+                }
+
+                if query.buy_token == BUY_ETH_ADDRESS {
+                    let sanitized_query = Query {
+                        buy_token: self.native_token,
+                        ..*query
+                    };
+
+                    tracing::debug!(
+                        ?query,
+                        ?sanitized_query,
+                        "estimate price for native token instead of ETH"
```

Comment:
This has become a bit confusing for me. I was under the impression we were referring to native token as the "native asset" of the network but apparently, throughout this repo it refers to the wrapped/erc20 version of this.

In any case, I suppose it makes sense then for us all to be aware that `native token = Wrapped(native asset)` for terminology. 

Here then I might suggest we change the comment to something like

```suggestion
                        "estimate price for wrapped native asset"
```

So not to hard-code ETH (since ETH isn't the native asset of all networks). For example
```
Gnosis Chain -> xDAI
Polygon -> MATIC
BSC -> BNB
```

Reply:
This comment you highlighted in the code seems inconsistent with the terminology we're using. On some places we are referring to native token as native asset like here.